core/contracts/financial-templates/implementation/demo/DepositBox.sol

```diff
@@ -0,0 +1,346 @@
+        OracleInterface oracle = _getOracle();
+        oracle.requestPrice(priceIdentifier, requestedTime);
+    }
+
+    // Ensure individual and global consistency when increasing collateral balances. Returns the change to the position.
+    function _incrementCollateralBalances(
+        DepositBoxData storage depositBoxData,
+        FixedPoint.Unsigned memory collateralAmount
+    ) internal returns (FixedPoint.Unsigned memory) {
+        _addCollateral(depositBoxData.rawCollateral, collateralAmount);
+        return _addCollateral(rawTotalDepositBoxCollateral, collateralAmount);
+    }
+
+    // Ensure individual and global consistency when decrementing collateral balances. Returns the change to the
+    // position. We elect to return the amount that the global collateral is decreased by, rather than the individual
+    // position's collateral, because we need to maintain the invariant that the global collateral is always
+    // <= the collateral owned by the contract to avoid reverts on withdrawals. The amount returned = amount withdrawn.
+    function _decrementCollateralBalances(
+        DepositBoxData storage depositBoxData,
+        FixedPoint.Unsigned memory collateralAmount
+    ) internal returns (FixedPoint.Unsigned memory) {
+        _removeCollateral(depositBoxData.rawCollateral, collateralAmount);
+        return _removeCollateral(rawTotalDepositBoxCollateral, collateralAmount);
+    }
+
+    function _resetWithdrawalRequest(DepositBoxData storage depositBoxData) internal {
+        depositBoxData.withdrawalRequestAmount = FixedPoint.fromUnscaledUint(0);
+        depositBoxData.requestPassTimestamp = 0;
+    }
+
+    function _depositBoxHasNoPendingWithdrawal(address user) internal view {
+        require(depositBoxes[user].requestPassTimestamp == 0, "Pending withdrawal");
+    }
+
+    function _getIdentifierWhitelist() internal view returns (IdentifierWhitelistInterface) {
+        return IdentifierWhitelistInterface(finder.getImplementationAddress(OracleInterfaces.IdentifierWhitelist));
+    }
+
+    function _getOracle() internal view returns (OracleInterface) {
+        return OracleInterface(finder.getImplementationAddress(OracleInterfaces.Oracle));
+    }
+
+    // Fetches a resolved Oracle price from the Oracle. Reverts if the Oracle hasn't resolved for this request.
+    function _getOraclePrice(uint256 requestedTime) internal view returns (FixedPoint.Unsigned memory) {
+        OracleInterface oracle = _getOracle();
+        require(oracle.hasPrice(priceIdentifier, requestedTime), "Unresolved oracle price");
+        int256 oraclePrice = oracle.getPrice(priceIdentifier, requestedTime);
+
+        // For simplicity we don't want to deal with negative prices.
+        if (oraclePrice < 0) {
+            oraclePrice = 0;
+        }
+        return FixedPoint.Unsigned(uint256(oraclePrice));
+    }
+
+    function _pfc() internal virtual override view returns (FixedPoint.Unsigned memory) {
```

Comment:
this function is super complex with a lot of nuanced stuff within it. Consider adding a comment like you did for the `_decrementCollateralBalances`